link/uprobe.go

```diff
@@ -141,28 +166,28 @@ func (ex *Executable) Uretprobe(symbol string, prog *ebpf.Program) (Link, error)
 
 // uprobe opens a perf event for the given binary/symbol and attaches prog to it.
 // If ret is true, create a uretprobe.
-func (ex *Executable) uprobe(symbol string, prog *ebpf.Program, ret bool) (*perfEvent, error) {
+func (ex *Executable) uprobe(symbol string, prog *ebpf.Program, opts *UprobeOptions, ret bool) (*perfEvent, error) {
 	if prog == nil {
 		return nil, fmt.Errorf("prog cannot be nil: %w", errInvalidInput)
 	}
 	if prog.Type() != ebpf.Kprobe {
 		return nil, fmt.Errorf("eBPF program type %s is not Kprobe: %w", prog.Type(), errInvalidInput)
 	}
 
-	sym, err := ex.symbol(symbol)
-	if err != nil {
-		return nil, fmt.Errorf("symbol '%s' not found in '%s': %w", symbol, ex.path, err)
-	}
-
-	// Symbols with location 0 from section undef are shared library calls and
-	// are relocated before the binary is executed. Dynamic linking is not
-	// implemented by the library, so mark this as unsupported for now.
-	if sym.Section == elf.SHN_UNDEF && sym.Value == 0 {
-		return nil, fmt.Errorf("cannot resolve %s library call '%s': %w", ex.path, symbol, ErrNotSupported)
+	var offset uint64
+	if opts != nil {
+		offset = opts.Offset
```

Comment:
This means that if we extend UprobeOptions in the future users will have to provide an offset, e.g:

```
Uprobe("main", prog, &UprobeOptions{Bar: true})  // doesn't work because we assume offset to be 0
```

The way around this would be to define `UprobeOptions.Offset == 0` to mean "not set", but then you can't trace on offset 0. Which is probably OK?

Reply:
So when would the probe be triggered for `offset 0`? When some buggy program tries to jump to a NULL function pointer?

I'm not a big fan of the `*uint64`. I prefer a separate `UprobeOffset()`, maybe without even introducing `UprobeOptions`? Downside: we also need `UretprobeOffset`.

Reply:
That's why I emphasized "in theory" :D. I don't know if there are legit use cases for `offset 0`.

`UprobeOffset()` sounds good (I'll evaluate the pid option first)

Reply:
> (I'll evaluate the pid option first)

To me pid and offset seem orthogonal, what am I missing?

Reply:
If `offset=0` the parsed offset will be valid so the offset will be correct, there is no need to use a pointer or `UprobeOffset`. (we already clarified the last phrase on Slack, I'm resolving the discussion)